rfd/0169-auto-updates-linux-agents.md

```diff
@@ -0,0 +1,934 @@
+
+This will require tracking the desired version of groups in the backend, which will add additional complexity to the rollout logic.
+
+## Details - Linux Agents
+
+We will ship a new auto-updater package for Linux servers written in Go that does not interface with the system package manager.
+It will be distributed as a separate package from Teleport, and manage the installation of the correct Teleport agent version manually.
+It will read the unauthenticated `/v1/webapi/find` endpoint from the Teleport proxy, parse new fields on that endpoint, and install the specified agent version according to the specified update plan.
+It will download the correct version of Teleport as a tarball, unpack it in `/var/lib/teleport`, and ensure it is symlinked from `/usr/local/bin`.
+
+Source code for the updater will live in the main Teleport repository, with the updater binary built from `tools/teleport-update`.
+
+### Installation
+
+```shell
+$ apt-get install teleport-ent-updater
+$ teleport-update enable --proxy example.teleport.sh
+
+# if not enabled already, configure teleport and:
+$ systemctl enable teleport
+```
+
+For grouped updates, a group identifier may be configured:
+```shell
+$ teleport-update enable --proxy example.teleport.sh --group staging
+```
+
+For air-gapped Teleport installs, the agent may be configured with a custom tarball path template:
+```shell
+$ teleport-update enable --proxy example.teleport.sh --template 'https://example.com/teleport-{{ .Edition }}-{{ .Version }}-{{ .Arch }}.tgz'
+```
+(Checksum will use template path + `.sha256`)
+
+### Filesystem
+
+```
+$ tree /var/lib/teleport
+/var/lib/teleport
+└── versions
+   ├── 15.0.0
+   │  ├── bin
+   │  │  ├── tsh
+   │  │  ├── tbot
+   │  │  ├── ... # other binaries
+   │  │  ├── teleport-update
+   │  │  └── teleport
+   │  ├── etc
+   │  │  └── systemd
+   │  │     └── teleport.service
+   │  └── backup
+   │     ├── sqlite.db
+   │     └── backup.yaml
+   ├── 15.1.1
+   │  ├── bin
+   │  │  ├── tsh
+   │  │  ├── tbot
+   │  │  ├── ... # other binaries
+   │  │  ├── teleport-update
+   │  │  └── teleport
+   │  └── etc
+   │     └── systemd
+   │        └── teleport.service
+   └── updates.yaml
+$ ls -l /usr/local/bin/tsh
+/usr/local/bin/tsh -> /var/lib/teleport/versions/15.0.0/bin/tsh
+$ ls -l /usr/local/bin/tbot
+/usr/local/bin/tbot -> /var/lib/teleport/versions/15.0.0/bin/tbot
+$ ls -l /usr/local/bin/teleport
+/usr/local/bin/teleport -> /var/lib/teleport/versions/15.0.0/bin/teleport
+$ ls -l /usr/local/bin/teleport-update
+/usr/local/bin/teleport-update -> /var/lib/teleport/versions/15.0.0/bin/teleport-update
+$ ls -l /usr/local/lib/systemd/system/teleport.service
+/usr/local/lib/systemd/system/teleport.service -> /var/lib/teleport/versions/15.0.0/etc/systemd/teleport.service
+```
+
+updates.yaml:
+```
+version: v1
+kind: agent_versions
```

Comment:
This looks an awful lot like a teleport resource, but it seems this is just a configuration file for the updater? If so, I might not use the same YAML structure with a version and kind to avoid people thinking that this is a Teleport resource that they need to `tctl create`. (Similar to how our access plugins or event handler are configured)

Reply:
@russjones requested this format explicitly

Reply:
https://github.com/gravitational/teleport/pull/40190#discussion_r1552493754